server/src/main/java/com/linecorp/centraldogma/server/internal/storage/DirectoryBasedStorageManager.java

```diff
@@ -347,11 +428,51 @@ private static boolean isValidChildName(String name) {
             return false;
         }
 
-        return !name.endsWith(SUFFIX_REMOVED);
+        return !name.endsWith(SUFFIX_REMOVED) && !name.endsWith(SUFFIX_PURGED);
     }
 
     @Override
     public String toString() {
         return Util.simpleTypeName(getClass()) + '(' + rootDir + ')';
     }
+
+    private final class StorageRemovalManager {
+
+        private static final String REMOVAL_TIMESTAMP_NAME = "removal.timestamp";
+
+        void mark(File file) {
+            final File removal = new File(file, REMOVAL_TIMESTAMP_NAME);
+            final String timestamp = DateTimeFormatter.ISO_INSTANT.format(Instant.now());
+            try {
+                Files.write(removal.toPath(), timestamp.getBytes(StandardCharsets.UTF_8));
+            } catch (IOException e) {
+                throw new StorageException(
+                        "failed to write a removal timestamp for " + childTypeName + ": " + removal);
+            }
+        }
+
+        void unmark(File file) {
+            final File removal = new File(file, REMOVAL_TIMESTAMP_NAME);
+            if (removal.exists()) {
+                if (!removal.delete()) {
+                    logger.warn("Failed to delete a removal timestamp for {}: {}", childTypeName, removal);
+                }
+            }
+        }
+
+        Instant readRemoval(File file) {
+            final File removal = new File(file, REMOVAL_TIMESTAMP_NAME);
+            if (!removal.exists()) {
+                return Instant.ofEpochMilli(file.lastModified());
```

Comment:
If a file has `.removed` postfix but no `REMOVAL_TIMESTAMP`, we determine the removal time by reading `lastModified()` to keep backward compatibility because old removed files don't have `REMOVAL_TIMESTAMP` before deploying this new purge logic.